### PR TITLE
SQL & php Spam performance changes for 20240308.1

### PR DESCRIPTION
**Pull request by joeiacoponi1**
[I do not appear to have write access to My-Little-Forum, so I created a fork and a pull request off it.]

4 total includes files modified: main, entry, index, thread.  All changes relate to queries using display_spam_query_(and or where).

Two  basic assumptions made:
1) All query results in base 20240308.1 are correct.  
2) The great majority (if not all) of MLF forums will have a far greater number of spam = 0 entries in akismet and b8 than spam = 1 entries.

**Changes**
For the code changes, given assumption 2 above (far fewer spam = 1 entries), I placed the retrieval of akismet and b8 records within a SQL subquery and only retrieved those where spam = 1. I then joined those results with the base query and changed the JOIN parameters depending on whether show_spam is set and whether we are returning individual entries or entire threads.

While performing the SQL changes, I noticed that the same basic SQL queries are repeated in a few instances, only differing based on if statement results, with the individual SQL statements run within the if statement.  The modifications I made consolidate the code a fair bit by specifying the base SQL code to be used for all scenarios, and then only setting an additional variable within the if statement results.  I then moved the SQL statement to run outside / after the if statement.

**Testing**
I have completed testing across all variables affected within these queries:
a) User:  Logged in vs out
b) Categories: No Categories specified - display all, Categories specified - display all, Categories specified - display one
c) Spam: No Spam entries, Spam entries exist and (not logged in or session(show_spam) not set), Spam entries exist and session(show_spam) is true.

Tested pages / modes / actions were index, user, show_posts, thread, posting and admin. 

All observed results (threads, posts, latest postings, total postings, total threads) for these changes match the results in base 20240308.1. 

In all instances tested, page load time went from 10-15 seconds in base 20240308.1 down to sub 1 second with these modifications. I do see about a 40% increase in page load time (from 500ms to 800ms avg) over 2.4.24, but page load is still acceptably under 1 second, even within our forum of 500,000 posts / 40,000 threads.

**Missing Index Note**
Responding to Auge's question about the missing akismet indexes during the double upgrade, I did add these indexes to the DB manually early during my testing process.  The indexes did not materially affect performance of base 20240308.1, but did dramatically affect performance with these pull request modifications in place.

Joe

### CODE REVIEW

**Comment by loesler** [comment]
Thank you for effort! Just a question on handling the changes correctly: Do we have to reconstruct #705?

**Comment by joeiacoponi1** [comment]
> Thank you for effort! Just a question on handling the changes correctly: Do we have to reconstruct #705?

You're quite welcome, Micha.  It was a good exercise, and hopefully others can find value in it.  

I'm not positive about your question. I made my changes off the current master branch, from yesterday. It appears #705 was merged into master 3 weeks ago, so I think you're ok.  There is also nothing in my changes which should affect how #705 works, as I understand it.

Joe

**Comment by joeiacoponi1** [comment]
> Thank you for effort! Just a question on handling the changes correctly: Do we have to reconstruct #705?

You're quite welcome, Micha. It was a good exercise, and hopefully others can find value in it.

Good catch. My modifications were based off master from 3/22, and I see that #705 was committed on 3/25. Given that, yes we will need to merge the changes from this and #705 within includes/main.inc.php.

Joe

**Comment by joeiacoponi1** [comment]
I've now committed the #705 changes to includes/main.inc.php to my master branch, so the pull request should be correct.

Joe

**Comment by loesler** [comment]
What a service. Thank you. ;-)

**Comment by auge8472** [comment]
I'll review the changes today or tomorrow.

File: main.inc.php [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	if(isset($usersettings_cookie[3])) {
+    $usersettings['fold_threads'] = $usersettings_cookie[3] == 1 ? 1 : 0;
+	} else {
+		$usersettings['fold_threads'] = $settings['fold_threads'];
+	}
+	$usersettings['thread_display'] = isset($usersettings_cookie[4]) && $usersettings_cookie[4] == 1 ? 1 : 0;
+	$usersettings['page'] = 1;
+	$usersettings['category'] = 0;
+	$_SESSION[$settings['session_prefix'].'usersettings'] = $usersettings;
+	setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days']));
```
Comment (loesler) [inline main.inc.php:111]: cookie_options - Missing

setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], cookie_options(TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days'])));

File: main.inc.php [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+if (isset($_REQUEST['toggle_sidebar'])) {
+	if (empty($_SESSION[$settings['session_prefix'].'usersettings']['sidebar'])) $_SESSION[$settings['session_prefix'].'usersettings']['sidebar'] = 1;
+	else $_SESSION[$settings['session_prefix'].'usersettings']['sidebar'] = 0;
+	setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days']));
```
Comment (loesler) [inline main.inc.php:117]: cookie_options - Missing

`setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], cookie_options(TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days'])));`

**Comment by joeiacoponi1** [comment]
Thank you for catching those Micha. I had done a code compare / merge, but for some reason those two didn't get moved over.

File: main.inc.php [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+} else {
+	$spam_totals = " AS spam_list ON spam_list.id = " . $db_settings['forum_table'] . ".id WHERE spam_list.id IS NULL";
+}
+if ($categories == false) {
+	$total_postings_category = "";		
+} else {
+	// there are categories
+	$total_postings_category = " AND category IN (" . $category_ids_query . ")";		
+}
+$total_threads_sql = $total_threads_postings_body . $spam_totals . $total_postings_category . $threads_check_sql;
```
Comment (loesler) [inline main.inc.php:254]: I got an error at this lines, i.e.

> [18-Apr-2024 19:19:12 UTC] PHP Fatal error:  Uncaught TypeError: mysqli_free_result(): Argument #1 ($result) must be of type mysqli_result, string given in /forum/includes/main.inc.php:257
> Stack trace:
> #0 /forum/includes/main.inc.php(257): mysqli_free_result('220')
> #1 /forum/index.php(36): include('/forum/...')
> #2 {main}
>   thrown in /forum/includes/main.inc.php on line 257

Suggestion (just using a different variable):
```
$total_threads_sql = $total_threads_postings_body . $spam_totals . $total_postings_category . $threads_check_sql;
$count_result = mysqli_query($connid, $total_threads_sql);
list($total_threads) = mysqli_fetch_row($count_result);
mysqli_free_result($count_result);

$total_postings_sql = $total_threads_postings_body . $spam_totals . $total_postings_category;
$count_result = mysqli_query($connid, $total_postings_sql);
list($total_postings) = mysqli_fetch_row($count_result);
mysqli_free_result($count_result);
```

Comment (joeiacoponi1) [reply main.inc.php:254]: Interesting that it worked fine on our site, but I see the issue. Let me fix and commit.

Comment (loesler) [reply main.inc.php:254]: My PHP version was 8.2 (I now check version 8.3). Do we use the same version? In any case, it is not a show stopper and does not overshadow your overall effort! So, every thing is fine and no further issue occurs so far. ;-)

Comment (joeiacoponi1) [reply main.inc.php:254]: Ahh, I just checked.  We're only running php 7.4.

Comment (loesler) [reply main.inc.php:254]: Maybe this is the reason (but I'm not an expert.)

Comment (joeiacoponi1) [reply main.inc.php:254]: That was my guess too, and the same solution.  It is now committed.

Comment (joeiacoponi1) [reply main.inc.php:254]: Ahh, now I remember why we’re on php 7.4! MyLittleForum 2.4.x is not compatible with php 8.x, so we have to stay on php 7.x until upgrade time.

**Review by auge8472 — CHANGES_REQUESTED** [review]
Why did you add the `main.inc.php` outside the directory `includes`? This must be removed.

**[edit]:** corrected a typo

**Comment by auge8472** [comment]
> I've now committed the #705 changes to includes/main.inc.php to my master branch, so the pull request should be correct.

This PR does **not** contain the changes of #705 (with stand of now (2024-04-19 14:24 CEST)). If you pulled the changes of #705 and merged it into your own master-branch and it is in a current state like our master-branch (plus your changes to this PR), you have to transmit those changes to your master on Github. If git reports a conflict, you have to push it with `git push -f origin master` (`-f` to force an overwrite of the remote repo).

**Comment by joeiacoponi1** [comment]
> Why did you add the `main.inc.php` outside the directory `includes`? This must be removed.
> 
> **[edit]:** corrected a typo

> This PR does **not** contain the changes of #705 (with stand of now (2024-04-19 14:24 CEST)). If you pulled the changes of #705 and merged it into your own master-branch and it is in a current state like our master-branch (plus your changes to this PR), you have to transmit those changes to your master on Github. If git reports a conflict, you have to push it with `git push -f origin master` (`-f` to force an overwrite of the remote repo).

Sorry about that, a little rusty with GitHub.  I've now moved main.inc.php from root back into /includes, and this does have the #705 changes along with all my changes from what the 4th commit shows.

Hopefully that cleans it all up and thanks for your assistance guys.

Joe

File: includes/main.inc.php [hunk 4]
```diff
@@ -108,13 +108,13 @@
 	$usersettings['page'] = 1;
 	$usersettings['category'] = 0;
 	$_SESSION[$settings['session_prefix'].'usersettings'] = $usersettings;
-	setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], cookie_options(TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days'])));
+	setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days']));
```
Comment (loesler) [inline includes/main.inc.php:111]: cookie_options is missing

File: includes/main.inc.php [hunk 5]
```diff
@@ -108,13 +108,13 @@
...
 	if (empty($_SESSION[$settings['session_prefix'].'usersettings']['sidebar'])) $_SESSION[$settings['session_prefix'].'usersettings']['sidebar'] = 1;
 	else $_SESSION[$settings['session_prefix'].'usersettings']['sidebar'] = 0;
-	setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], cookie_options(TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days'])));
+	setcookie($settings['session_prefix'].'usersettings', $_SESSION[$settings['session_prefix'].'usersettings']['user_view'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_order'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['sidebar'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['fold_threads'].'.'.$_SESSION[$settings['session_prefix'].'usersettings']['thread_display'], TIMESTAMP + (3600 * 24 * $settings['cookie_validity_days']));
```
Comment (loesler) [inline includes/main.inc.php:117]: cookie_options is missing

Comment (joeiacoponi1) [reply includes/main.inc.php]: OK, give that one more try and thanks for hanging in there.

Comment (loesler) [reply includes/main.inc.php]: Thank you for fixing!

Comment (joeiacoponi1) [reply includes/main.inc.php]: Gladly, and sorry it was so painful.

Comment (loesler) [reply includes/main.inc.php]: From my side the changes look fine and it works in my forum. Maybe @auge8472 has a few more comments on the changes.

Comment (auge8472) [reply includes/main.inc.php]: I began to compare the queries yesterday but I'm not done. Please be patient.

File: includes/entry.inc.php [hunk 6]
```diff
@@ -134,22 +134,17 @@
...
+	if ($entrydata['spam'] == 1 && isset($id)) {
+		$spam_sql_and .= " OR `ft`.`id` = " . intval($id);
+	}
+	$entry_sql = 
+		"SELECT ft.id, ft.pid, ft.tid, ft.user_id, UNIX_TIMESTAMP(ft.time) AS time, UNIX_TIMESTAMP(ft.time + INTERVAL " . $time_difference . " MINUTE) AS disp_time, UNIX_TIMESTAMP(last_reply) AS last_reply, name, user_name, user_type, subject, category, marked, text, rst.user_id AS req_user, " . $db_settings['akismet_rating_table'] . ".spam AS akismet_spam, " . $db_settings['b8_rating_table'] . ".spam AS b8_spam, " . $db_settings['akismet_rating_table'] . ".spam_check_status AS akismet_checked, " . $db_settings['b8_rating_table'] . ".training_type AS b8_checked 
+		FROM " . $db_settings['forum_table'] . " AS ft LEFT JOIN " . $db_settings['userdata_table'] . " ON " . $db_settings['userdata_table'] . ".user_id = ft.user_id LEFT JOIN " . $db_settings['read_status_table'] . " AS rst ON rst.posting_id = ft.id AND rst.user_id = " . intval($tmp_user_id) . " LEFT JOIN " . $db_settings['akismet_rating_table'] . " ON " . $db_settings['akismet_rating_table'] . ".eid = ft.id LEFT JOIN " . $db_settings['b8_rating_table'] . " ON " . $db_settings['b8_rating_table'] . ".eid = ft.id
```
Comment (auge8472) [inline includes/entry.inc.php:143]: One question: is it really necessary to join the akismet-rating and b8-rating tables individually and to join them again in the next line as a unionised subquery with the same criteria?

File: includes/main.inc.php [hunk 7]
```diff
@@ -221,36 +221,45 @@
 }
 
 // count spam:
-$count_spam_result = mysqli_query($connid, "SELECT COUNT(*) FROM " . $db_settings['forum_table'] . " LEFT JOIN " . $db_settings['akismet_rating_table'] . " ON " . $db_settings['akismet_rating_table'] . ".eid = " . $db_settings['forum_table'] . ".id LEFT JOIN " . $db_settings['b8_rating_table'] . " ON " . $db_settings['b8_rating_table'] . ".eid = " . $db_settings['forum_table'] . ".id WHERE (" . $db_settings['akismet_rating_table'] . ".spam = 1 OR " . $db_settings['b8_rating_table'] . ".spam = 1)");
+$count_spam_sql = 
+	"SELECT COUNT(*) FROM " . $db_settings['forum_table'] . " INNER JOIN (SELECT eid FROM " . $db_settings['akismet_rating_table'] . " WHERE " . $db_settings['akismet_rating_table'] . ".spam = 1 UNION SELECT eid FROM " . $db_settings['b8_rating_table'] . "	WHERE " . $db_settings['b8_rating_table'] . ".spam = 1) spam_list ON spam_list.eid = " . $db_settings['forum_table'] . ".id";
```
Comment (auge8472) [inline includes/main.inc.php:225]: I know, it's not necessary but the query in entry.inc.php uses the syntax `JOIN (SELECT …) AS foo ON …`, this one not. ***I*** personally prefer the keyword `AS` for better readability. At least we should handle it in a uniform way.

File: includes/index.inc.php [hunk 8]
```diff
@@ -137,40 +159,22 @@
...
-				LEFT JOIN ".$db_settings['read_status_table']." AS rst ON rst.posting_id = ft.id AND rst.user_id = ". intval($tmp_user_id) ."
-				LEFT JOIN " . $db_settings['akismet_rating_table'] . " ON " . $db_settings['akismet_rating_table'] . ".`eid` = `ft`.`id` 
-				LEFT JOIN " . $db_settings['b8_rating_table'] . " ON " . $db_settings['b8_rating_table'] . ".`eid` = `ft`.`id` 
-				WHERE (" . $db_settings['akismet_rating_table'] . ".spam = 0 AND " . $db_settings['b8_rating_table'] . ".spam = 0) 
-				AND category IN (". $category_ids_query .")
-				ORDER BY ft.time DESC LIMIT ".$settings['latest_postings']) or raise_error('database_error', mysqli_error($connid));
+			$latest_postings_category_sql = " AND category IN (". $category_ids_query .")";		
 		}
 	}
+	$latest_postings_sql = $latest_postings_body_sql . $latest_postings_category_sql . $latest_postings_order_by_sql ;
```
Comment (auge8472) [inline includes/index.inc.php:176]: Nitpicking: please remove the blank in front of the trailing semicolon.

File: includes/thread.inc.php [hunk 9]
```diff
@@ -66,23 +68,32 @@
...
+        subject, hp, location, ip, text, cache_text, show_signature, views, category, locked, ip,
+        user_name, user_type, user_email, email_contact, user_hp, user_location, signature, cache_signature, edit_key, rst.user_id AS req_user,
+        " . $db_settings['akismet_rating_table'] . ".spam AS akismet_spam, spam_check_status,
+        " . $db_settings['b8_rating_table'] . ".spam AS b8_spam, training_type
+        FROM " . $db_settings['forum_table'] . " AS ft
+        LEFT JOIN " . $db_settings['entry_cache_table'] . " ON " . $db_settings['entry_cache_table'] . ".cache_id=ft.id
+        LEFT JOIN " . $db_settings['userdata_table'] . " ON " . $db_settings['userdata_table'] . ".user_id=ft.user_id
+        LEFT JOIN " . $db_settings['userdata_cache_table'] . " ON " . $db_settings['userdata_cache_table'] . ".cache_id=" . $db_settings['userdata_table'] . ".user_id
+        LEFT JOIN " . $db_settings['read_status_table'] . " AS rst ON rst.posting_id = ft.id AND rst.user_id = " . intval($tmp_user_id) . "
+		LEFT JOIN " . $db_settings['akismet_rating_table'] . " ON " . $db_settings['akismet_rating_table'] . ".eid = `ft`.id
```
Comment (auge8472) [inline includes/thread.inc.php:92]: Same question as for entry.inc.php. Is it really necessary to perform this and the following join when the next one does the same with the union subquery?

File: includes/entry.inc.php (hunk 6 above)
Comment (joeiacoponi1) [reply includes/entry.inc.php:143]: Ahh, very good question. My simple answer is yes, and even though it's a bit weird looking, it is extremely performant. I'll explain my approach and maybe we can find cleaner solution.

1) The queries need to be performant.
2) Through my testing, creating the subquery which retrieves only the spam = 1 records is always far more performant than any method which retrieves spam = 0 records.
3) I wanted all of the queries I modified to be as consistent in their logic as possible, as long as #1 above is satisfied.
4) The basic challenge is that these queries for entry.inc.php and thread.inc.php require additional fields that the other modified queries don't (akismet_spam, b8_spam, akismet_checked, b8_checked). Retrieving these fields within the inner query UNION could create multiple records for a single entry, as it is quite possible the spam values are different in the two tables. Plus, we need to individually retrieve the checked status for both.
5) As an alternative, I tried to retrieve all required fields within the subquery using a SELECT on mlf2_entries WITH LEFT JOINs on akismet and b8 and WHERE akismet spam = 1 OR b8 spam = 1.  This resulted in another slow performing query, so I abandoned that approach.  
6) Assuming that the subquery as currently modified always retrieves a limited number of results, adding the JOIN to akismet and b8 in the outer query to get the additional fields only retrieves a few records, and they are retrieved via index. This makes it an efficient approach.

I'm always open to looking at a different approach, but this was the result that was performant across all queries and retrieved the same results as 20240308.1.

Note: 
I'll also take a look at your other comments and make the appropriate changes.  I completely agree about handling these queries in a uniform way, and consistent with what is already in production.  In fact when making my changes I made a point to clean up some of the sql code from 20240308.1 to do just that.

Joe

Comment (auge8472) [reply includes/entry.inc.php:143]: Thank you very much for your detailed explanation.

File: main.inc.php (hunk 2 above)
Comment (auge8472) [reply main.inc.php]: It looks to me, like every occurence of missing `cookie_options` is solved. Can we mark *this part* as resolved?

Comment (loesler) [reply main.inc.php]: Yes, I think so. Go ahead.

File: includes/main.inc.php (hunk 7 above)
Comment (joeiacoponi1) [reply includes/main.inc.php]: I've now updated these two items along with a few other code formatting items I'd come across, to match consistency with existing code.

**Comment by auge8472** [comment]
As far as I can see, there is nothing left. IMHO @joeiacoponi1 solved all requests. Nevertheless, Github believes that another change is still pending. @loesler, @joeiacoponi1 Please take a look and tell me please what I am missing.

If there's nothing, this is ready to be squashed. 😀👍🎉

**Review by auge8472 — APPROVED** [review]

**Comment by auge8472** [comment]
... aaaaand done. 🎉

**Comment by loesler** [comment]
Good work!

**Pull request merged**